components/script/task_queue.rs

```diff
@@ -53,22 +58,85 @@ impl<T: QueuedTaskConversion> TaskQueue<T> {
+    fn process_message(
+        &self,
+        msg: T,
+        currently_inactive: &HashSet<PipelineId>,
+        incoming: &mut Vec<T>,
+    ) {
+        if msg.is_wake_up() {
+            return;
+        }
+        let mut inactive = self.inactive.borrow_mut();
+        // Hold back tasks for currently not fully-active documents.
+        // https://html.spec.whatwg.org/multipage/#event-loop-processing-model:fully-active
+        if let Some(pipeline_id) = msg.pipeline_id() {
+            if currently_inactive.contains(pipeline_id) {
+                let inactive_queue = inactive
+                    .entry(pipeline_id.clone())
+                    .or_insert(VecDeque::new());
+                inactive_queue
+                    .push_back(msg.into_queued_task().expect(
+                        "Incoming messages should always be convertible into queued tasks",
+                    ));
+                if incoming.is_empty() {
+                    // Ensure there is at least one message.
+                    // Otherwise if the just stored inactive message
+                    // was the first and last of this iteration,
+                    // it will result in a spurious wake-up of the event-loop.
+                    incoming.push(T::inactive_msg());
+                }
+                return;
+            }
+        }
+        incoming.push(msg);
+    }
+
+    /// Release previously held-back tasks for documents that are now fully-active.
+    fn release_tasks_for_fully_active_documents(
+        &self,
+        currently_inactive: &HashSet<PipelineId>,
+        incoming: &mut Vec<T>,
+    ) {
+        let previously_inactive: HashSet<PipelineId> = {
+            let inactive = self.inactive.borrow();
+            inactive.keys().map(|id| id.clone()).collect()
+        };
+        let mut inactive = self.inactive.borrow_mut();
+        for pipeline_id in previously_inactive.difference(&currently_inactive) {
+            let (_, inactive_queue) = inactive
+                .remove_entry(&pipeline_id)
+                .expect("Inactive should contain this key");
+            for queued_task in inactive_queue {
+                incoming.push(T::from_queued_task(queued_task));
+            }
         }
     }
 
     /// Process incoming tasks, immediately sending priority ones downstream,
     /// and categorizing potential throttles.
-    fn process_incoming_tasks(&self, first_msg: T) {
-        let mut incoming = Vec::with_capacity(self.port.len() + 1);
-        if !first_msg.is_wake_up() {
-            incoming.push(first_msg);
-        }
+    fn process_incoming_tasks(&self, first_msg: T, currently_inactive: &HashSet<PipelineId>) {
+        let mut incoming = vec![];
+
+        // 1. Make any previously stored task from now fully-active document available.
+        self.release_tasks_for_fully_active_documents(currently_inactive, &mut incoming);
```

Comment:
This seems to be the only use of `release_tasks_for_fully_active_documents`, is there any reason it's passed a mutable `Vec<_>` instead of returning one?